File: src/nodejs/src/repositories/CachedOrderRepository.ts
import { CacheEntryOptions, defaultCacheOptions } from "../config/CacheEntryOptions";
import { IOrderRepository } from "./IOrderRepository";
import { JsonUtils } from "../utils/jsonUtils";
import { Order } from "../entities/Order";
import logger from "../utils/logger";
import redisClient from "../utils/redisClient";

/**
 * Decorator for IOrderRepository that adds Redis caching functionality.
 * Implements the Decorator pattern to wrap an existing repository with caching capabilities.
 */
export class CachedOrderRepository implements IOrderRepository {
  /**
   * Maximum number of retry attempts for cache operations.
   */
  private readonly maxRetries = 3;

  /**
   * Creates a new instance of CachedOrderRepository.
   * @param innerRepository - The repository to decorate with caching.
   * @param cacheOptions - Optional cache configuration options.
   */
  constructor(
    private innerRepository: IOrderRepository,
    private cacheOptions: CacheEntryOptions = defaultCacheOptions
  ) {
    this.innerRepository = innerRepository;
    this.cacheOptions = cacheOptions;
  }

  /**
   * Retrieves an order by its ID, using cache when available.
   * @param id - The order ID.
   * @returns The order if found, null otherwise.
   */
  async getOrderAsync(id: number): Promise<Order | null> {
    const key = this.getOrderCacheKey(id);

    try {
      const cachedData = await redisClient.get(key);

      if (cachedData) {
        logger.debug(`Cache hit for order: ${id}`);
        const data = JsonUtils.deserialize<Order>(cachedData);
        if (data) {
          return Object.assign(new Order(), data);
        }
        return null;
      }

      logger.debug(`Cache miss for order: ${id}`);
    } catch (error) {
      logger.warn(`Failed to read from cache for order ${id}:`, error);
    }

    const order = await this.innerRepository.getOrderAsync(id);

    if (order) {
      await this.saveToCacheAsync(order);
    }

    return order;
  }

  /**
   * Adds a new order to the repository.
   * @param order - The order to add.
   * @returns The added order with generated ID.
   */
  async addAsync(order: Order): Promise<Order> {
    const createdOrder = await this.innerRepository.addAsync(order);
    await this.removeFromCacheAsync(undefined, createdOrder.patientId);
    return createdOrder;
  }

  /**
   * Updates an order's message.
   * @param id - The order ID.
   * @param message - The new message.
   * @param updatedAt - The update timestamp.
   * @returns The updated order if found, null otherwise.
   */
  async updateAsync(id: number, message: string, updatedAt: Date): Promise<Order | null> {
    const updatedOrder = await this.innerRepository.updateAsync(id, message, updatedAt);

    if (updatedOrder) {
      await this.removeFromCacheAsync(id, updatedOrder.patientId);
    }

    return updatedOrder;
  }

  /**
   * Saves an order to the cache with retry logic.
   * @param order - The order to cache.
   */
  private async saveToCacheAsync(order: Order): Promise<void> {
    const key = this.getOrderCacheKey(order.id);
    const value = JSON.stringify(order);
    const ttl = this.getTtlInSeconds();

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        if (ttl > 0) {
          await redisClient.setex(key, ttl, value);
        } else {
          await redisClient.set(key, value);
        }
        logger.debug(`Cached order: ${order.id}`);
        return;
      } catch (error) {
        const delay = Math.pow(2, attempt) * 1000 + Math.random() * 1000;
        logger.warn(
          `Cache write failed for order ${order.id}. Attempt ${attempt}/${this.maxRetries}. Retrying in ${delay}ms`,
          error
        );

        if (attempt < this.maxRetries) {
          await this.delayAsync(delay);
        } else {
          logger.error(`Failed to cache order ${order.id} after ${this.maxRetries} attempts`, error);
        }
      }
    }
  }

  /**
   * Removes cache entries related to an order.
   * @param orderId - The order ID to remove from cache (optional).
   * @param patientId - The patient ID to remove related cache (optional).
   */
  private async removeFromCacheAsync(orderId?: number, patientId?: number): Promise<void> {
    try {
      const keysToDelete: string[] = [];

      if (orderId !== undefined) {
        keysToDelete.push(this.getOrderCacheKey(orderId));
      }

      if (patientId !== undefined) {
        keysToDelete.push(this.getPatientCacheKey(patientId));
      }

      if (keysToDelete.length > 0) {
        await redisClient.del(...keysToDelete);
        logger.debug(`Removed cache keys: ${keysToDelete.join(", ")}`);
      }
    } catch (error) {
      logger.warn("Failed to remove cache entries:", error);
    }
  }

  /**
   * Gets the cache key for an order.
   * @param id - The order ID.
   * @returns The cache key.
   */
  private getOrderCacheKey(id: number): string {
    return `order:${id}`;
  }

  /**
   * Gets the cache key for a patient.
   * @param patientId - The patient ID.
   * @returns The cache key.
   */
  private getPatientCacheKey(patientId: number): string {
    return `patient:${patientId}`;
  }

  /**
   * Calculates the TTL (Time To Live) in seconds based on cache options.
   * @returns The TTL in seconds.
   */
  private getTtlInSeconds(): number {
    const { slidingExpirationMinutes, absoluteExpirationRelativeToNowMinutes } = this.cacheOptions;

    // Use the smaller of the two expiration times if both are set
    if (slidingExpirationMinutes && absoluteExpirationRelativeToNowMinutes) {
      return Math.min(slidingExpirationMinutes, absoluteExpirationRelativeToNowMinutes) * 60;
    }

    if (slidingExpirationMinutes) {
      return slidingExpirationMinutes * 60;
    }

    if (absoluteExpirationRelativeToNowMinutes) {
      return absoluteExpirationRelativeToNowMinutes * 60;
    }

    return 0;
  }

  /**
   * Sleep for a specified duration.
   * @param ms - Duration in milliseconds.
   */
  private delayAsync(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: src/python/app/repositories/caches/cached_order_repository.py
"""Cached order repository with Redis decorator."""

import asyncio
import json
import logging
from datetime import datetime
from typing import Optional
import redis.asyncio as redis
from app.entities import PatientOrder
from app.repositories.order_repository import OrderRepository
from app.configs.cache_config import cache_settings

logger = logging.getLogger(__name__)


class CachedOrderRepository:
    """Decorator for OrderRepository that adds Redis caching."""

    def __init__(self, inner_repository: OrderRepository, redis_client: redis.Redis):
        """
        Initialize cached repository.

        Args:
            inner_repository: The inner order repository to which calls are delegated
            redis_client: Redis client for caching
        """
        self._inner_repository = inner_repository
        self._redis = redis_client
        self._ttl = cache_settings.cache_ttl_seconds

    @staticmethod
    def _get_order_cache_key(order_id: int) -> str:
        """Get cache key for an order."""
        return f"order:{order_id}"

    @staticmethod
    def _get_patient_cache_key(patient_id: int) -> str:
        """Get cache key for a patient."""
        return f"patient:{patient_id}"

    async def _save_to_cache(self, order: PatientOrder) -> None:
        """Save order to cache."""
        try:
            key = self._get_order_cache_key(order.id)
            order_dict = {
                "id": order.id,
                "message": order.message,
                "patient_id": order.patient_id,
                "created_at": order.created_at.isoformat(),
                "updated_at": order.updated_at.isoformat(),
            }
            await self._redis.setex(key, self._ttl, json.dumps(order_dict))
            logger.debug("Saved order %s to cache", order.id)
        except Exception as e:
            logger.error("Error occurred while saving order to cache with key %s after retries: %s", key, str(e))

    async def _remove_from_cache(self, order_id: int, patient_id: int) -> None:
        """Remove order and associated patient from cache."""
        try:
            tasks = []

            if order_id > 0:
                order_cache_key = self._get_order_cache_key(order_id)
                tasks.append(self._redis.delete(order_cache_key))

            if patient_id > 0:
                patient_cache_key = self._get_patient_cache_key(patient_id)
                tasks.append(self._redis.delete(patient_cache_key))

            if tasks:
                await asyncio.gather(*tasks)

            logger.debug("Removed cache for order %s and patient %s", order_id, patient_id)
        except Exception as e:
            logger.error(
                "Error occurred while removing cache for Order %s or Patient %s after retries: %s",
                order_id,
                patient_id,
                str(e),
            )

    async def get_order(self, order_id: int) -> Optional[PatientOrder]:
        """Get order by ID (with caching)."""
        key = self._get_order_cache_key(order_id)

        try:
            cached_data = await self._redis.get(key)
            if cached_data:
                logger.debug("Order %s retrieved from cache", order_id)
                order_dict = json.loads(cached_data)
                order = PatientOrder(
                    id=order_dict["id"],
                    message=order_dict["message"],
                    patient_id=order_dict["patient_id"],
                    created_at=datetime.fromisoformat(order_dict["created_at"]),
                    updated_at=datetime.fromisoformat(order_dict["updated_at"]),
                )
                return order
        except Exception as e:
            logger.warning("Failed to get order from cache: %s", str(e))

        order = await self._inner_repository.get_order(order_id)
        if order is not None:
            await self._save_to_cache(order)
        return order

    async def add(self, order: PatientOrder) -> PatientOrder:
        """Add new order (invalidates patient cache)."""
        created_order = await self._inner_repository.add(order)
        await self._remove_from_cache(0, created_order.patient_id)
        return created_order

    async def update_message(self, order: PatientOrder) -> PatientOrder:
        """Update order message (invalidates order and patient cache)."""
        updated_order = await self._inner_repository.update_message(order)
        await self._remove_from_cache(order.id, updated_order.patient_id)
        return updated_order

    async def update(self, order_id: int, message: str, updated_at: datetime) -> Optional[PatientOrder]:
        """Update order with new message and timestamp (invalidates order and patient cache)."""
        updated_order = await self._inner_repository.update(order_id, message, updated_at)

        if updated_order is None:
            return None

        await self._remove_from_cache(order_id, updated_order.patient_id)
        return updated_order
